service/graph.go

```diff
@@ -18,20 +18,44 @@ import (
 	"context"
 	"net/http"
 
+	"gonum.org/v1/gonum/graph/simple"
+	"gonum.org/v1/gonum/graph/topo"
+
 	"go.opentelemetry.io/collector/component"
 )
 
 var _ pipelines = (*pipelinesGraph)(nil)
 
-type pipelinesGraph struct{}
+type pipelinesGraph struct {
+	// All component instances represented as nodes, with directed edges indicating data flow.
+	componentGraph *simple.DirectedGraph
+}
 
 func (g *pipelinesGraph) StartAll(ctx context.Context, host component.Host) error {
-	// TODO actual implementation
+	nodes, err := topo.Sort(g.componentGraph)
+	if err != nil {
+		return err
+	}
+	// Start exporters first, and work towards receivers
+	for i := len(nodes) - 1; i >= 0; i-- {
+		if compErr := nodes[i].(component.Component).Start(ctx, host); compErr != nil {
+			return compErr
+		}
+	}
 	return nil
 }
 
 func (g *pipelinesGraph) ShutdownAll(ctx context.Context) error {
-	// TODO actual implementation
+	nodes, err := topo.Sort(g.componentGraph)
+	if err != nil {
+		return err
+	}
+	// Stop receivers first, and work towards exporters
+	for i := 0; i < len(nodes); i++ {
+		if compErr := nodes[i].(component.Component).Shutdown(ctx); compErr != nil {
```

Comment:
I think we should not stop if error, instead accumulate them using `multierr.Append` and continue.

Reply:
Makes sense. Fixed.